bam_sort.c

```diff
@@ -1098,58 +1128,95 @@ int bam_merge_core2(int by_qname, char* sort_tag, const char *out, const char *m
+            if (reg_idx == NULL) {
+                fprintf(stderr, "[%s] failed to load index for %s. Random alignment retrieval only works for indexed BAM or CRAM files.\n",
                         __func__, fn[i]);
+                free(rtrans);
+                rtrans = NULL;
                 goto fail;
             }
-            if (mapped_tid != INT32_MIN) {
-                iter[i] = sam_itr_queryi(idx, mapped_tid, beg, end);
+
+            int mapped_tid = INT32_MIN;
+            if (fn_bed) {
+                hts_reglist_t *rl = duplicate_reglist(lreg, nreg);
+                iter[i] = sam_itr_regions(reg_idx, hdr[i], rl, nreg);
             } else {
-                iter[i] = sam_itr_queryi(idx, HTS_IDX_NONE, 0, 0);
+                // (rtrans[i*n+tid]) Look up what hout tid translates to in input tid space
+                mapped_tid = rtrans[i*sam_hdr_nref(hout)+tid];
+                if (mapped_tid != INT32_MIN) {
+                    iter[i] = sam_itr_queryi(reg_idx, mapped_tid, beg, end);
+                } else {
+                    iter[i] = sam_itr_queryi(reg_idx, HTS_IDX_NONE, 0, 0);
+                }
             }
-            hts_idx_destroy(idx);
+
             if (iter[i] == NULL) {
-                if (mapped_tid != INT32_MIN) {
-                    fprintf(stderr,
-                            "[%s] failed to get iterator over "
-                            "{%s, %d, %"PRIhts_pos", %"PRIhts_pos"}\n",
-                            __func__, fn[i], mapped_tid, beg, end);
+                if (fn_bed) {
+                    fprintf(stderr, "[%s] failed to get multi-region iterator "
+                            "{%s, %s}\n", __func__, fn[i], fn_bed);
                 } else {
-                    fprintf(stderr,
-                            "[%s] failed to get iterator over "
-                            "{%s, HTS_IDX_NONE, 0, 0}\n",
-                            __func__, fn[i]);
+                    if (mapped_tid != INT32_MIN) {
+                        fprintf(stderr,
+                                "[%s] failed to get iterator over "
+                                "{%s, %d, %"PRIhts_pos", %"PRIhts_pos"}\n",
+                                __func__, fn[i], mapped_tid, beg, end);
+                    } else {
+                        fprintf(stderr,
+                                "[%s] failed to get iterator over "
+                                "{%s, HTS_IDX_NONE, 0, 0}\n",
+                                __func__, fn[i]);
+                    }
                 }
+                hts_idx_destroy(reg_idx);
+                free(rtrans);
+                rtrans = NULL;
                 goto fail;
             }
+
+            hts_idx_destroy(reg_idx);
         }
+
         free(rtrans);
         rtrans = NULL;
     } else {
         for (i = 0; i < n; ++i) {
-            if (hdr[i] == NULL) {
+            if (hts_get_format(fp[i])->format != sam) {
```

Comment:
It might be possible to simplify this.  Back in commit 3fdca5b6e5 (when this test was added) the input header was destroyed for non-SAM files before getting here, so it used `sam_itr_queryi(NULL, HTS_IDX_REST, 0, 0)` to avoid having to use `hdr[i]` when getting records.  Now `hdr[i]` still exists, so `sam_read1()` should work for everything and it should be possible to just say `iter[i] = NULL;` for all file types.